File: examples/Delphi/DSACmp1.pas
unit DSACmp1;

interface

uses
{$IFDEF WIN32}
  Windows,
  Messages,
  Graphics,
  Controls,
  Forms,
  Dialogs,
  StdCtrls,
  ComCtrls,
{$ENDIF}
{$IFDEF LINUX}
  QForms,
  QStdCtrls,
  QControls,
  QExtCtrls,
  QComCtrls,
{$ENDIF}
  SysUtils,
  Classes,
  LbCipher,
  LbClass,
  LbAsym,
  LbDSA;

type
  TForm1 = class(TForm)
    LbDSA1: TLbDSA;
    btnGenParams: TButton;
    mmoP: TMemo;
    Label1: TLabel;
    Label2: TLabel;
    mmoG: TMemo;
    Label3: TLabel;
    Label4: TLabel;
    mmoY: TMemo;
    Label5: TLabel;
    edtMsg: TEdit;
    Label6: TLabel;
    edtR: TEdit;
    edtS: TEdit;
    btnSign: TButton;
    Label7: TLabel;
    Label8: TLabel;
    edtQ: TEdit;
    edtX: TEdit;
    btnVerify: TButton;
    StatusBar1: TStatusBar;
    edtSeed: TEdit;
    Label9: TLabel;
    Label10: TLabel;
    edtXKey: TEdit;
    Label11: TLabel;
    edtKKey: TEdit;
    btnGenXY: TButton;
    btnSetParams: TButton;
    btnSetXY: TButton;
    chkAbort: TCheckBox;
    procedure btnGenParamsClick(Sender: TObject);
    procedure btnSignClick(Sender: TObject);
    procedure btnVerifyClick(Sender: TObject);
    procedure btnGenXYClick(Sender: TObject);
    procedure btnSetParamsClick(Sender: TObject);
    procedure btnSetXYClick(Sender: TObject);
    procedure LbDSA1GetSeed(Sender: TObject; var Block: TLbDSABlock);
    procedure LbDSA1GetXKey(Sender: TObject; var Block: TLbDSABlock);
    procedure LbDSA1GetKKey(Sender: TObject; var Block: TLbDSABlock);
    procedure LbDSA1GetR(Sender: TObject; var Block: TLbDSABlock);
    procedure LbDSA1GetS(Sender: TObject; var Block: TLbDSABlock);
    procedure LbDSA1Progress(Sender: TObject; var Abort: Boolean);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  Form1: TForm1;

implementation

{$R *.dfm}

uses
  LbUtils, LbRSA;


procedure TForm1.btnGenParamsClick(Sender: TObject);
  { generate DSA parameters p, q, and g }
begin
  Screen.Cursor := crHourGlass;
  StatusBar1.SimpleText := ' Generating DSA parameters - this may take a while';

  try
    if LBDSA1.GeneratePQG then begin
      edtQ.Text := LbDSA1.PrivateKey.QAsString;
      mmoP.Lines.Text := LbDSA1.PrivateKey.PAsString;
      mmoG.Lines.Text := LbDSA1.PrivateKey.GAsString;
      StatusBar1.SimpleText := '';
    end else
      StatusBar1.SimpleText := ' DSA parameter failure';
  finally
    Screen.Cursor := crDefault;
  end;
end;

procedure TForm1.btnGenXYClick(Sender: TObject);
  { generate key parameters x and y }
begin
  Screen.Cursor := crHourGlass;
  StatusBar1.SimpleText := ' Generating X and Y';

  try
    LbDSA1.GenerateXY;
    edtX.Text := LbDSA1.PrivateKey.XAsString;
    mmoY.Lines.Text := LbDSA1.PublicKey.YAsString;
    StatusBar1.SimpleText := '';
  finally
    Screen.Cursor := crDefault;
  end;
end;

procedure TForm1.btnSignClick(Sender: TObject);
  { sign message }
begin
  Screen.Cursor := crHourglass;
  StatusBar1.SimpleText := ' Signing message';

  try
    LbDSA1.SignString(edtMsg.Text);
    edtR.Text := LbDSA1.SignatureR.IntStr;
    edtS.Text := LbDSA1.SignatureS.IntStr;
    StatusBar1.SimpleText := '';
  finally
    Screen.Cursor := crDefault;
  end;
end;

procedure TForm1.btnVerifyClick(Sender: TObject);
  { verify signature }
begin
  Screen.Cursor := crHourglass;
  StatusBar1.SimpleText := ' Verifying signature';

  try
    if LbDSA1.VerifyString(edtMsg.Text) then
      StatusBar1.SimpleText := ' Verification PASSED'
    else
      StatusBar1.SimpleText := ' Verification FAILED';
  finally
    Screen.Cursor := crDefault;
  end;
end;

procedure TForm1.btnSetParamsClick(Sender: TObject);
  { assign DSA parameters p, q, and g }
begin
  LbDSA1.PrivateKey.QAsString := edtQ.Text;
  LbDSA1.PrivateKey.PAsString := mmoP.Lines.Text;
  LbDSA1.PrivateKey.GAsString := mmoG.Lines.Text;
  LbDSA1.PublicKey.QAsString := edtQ.Text;
  LbDSA1.PublicKey.PAsString := mmoP.Lines.Text;
  LbDSA1.PublicKey.GAsString := mmoG.Lines.Text;
end;

procedure TForm1.btnSetXYClick(Sender: TObject);
  { assign key parameters x and y }
begin
  LbDSA1.PrivateKey.XAsString := edtX.Text;
  LbDSA1.PublicKey.YAsString := mmoY.Text;
end;

procedure TForm1.LbDSA1GetSeed(Sender: TObject; var Block: TLbDSABlock);
  { return seed }
begin
  HexToBuffer(edtSeed.Text, Block, SizeOf(Block));
end;

procedure TForm1.LbDSA1GetXKey(Sender: TObject; var Block: TLbDSABlock);
  { return XKey }
begin
  HexToBuffer(edtXKey.Text, Block, SizeOf(Block));
end;

procedure TForm1.LbDSA1GetKKey(Sender: TObject; var Block: TLbDSABlock);
  { return YKey }
begin
  HexToBuffer(edtKKey.Text, Block, SizeOf(Block));
end;

procedure TForm1.LbDSA1GetR(Sender: TObject; var Block: TLbDSABlock);
  { return signature (r) }
begin
  HexToBuffer(edtR.Text, Block, SizeOf(Block));
end;

procedure TForm1.LbDSA1GetS(Sender: TObject; var Block: TLbDSABlock);
  { return signature (s) }
begin
  HexToBuffer(edtS.Text, Block, SizeOf(Block));
end;

procedure TForm1.LbDSA1Progress(Sender: TObject; var Abort: Boolean);
  { allow message loop to run }
begin
  Application.ProcessMessages;
  Abort := chkAbort.Checked;
end;

end.

File: examples/CBuilder/DSACmp1.cpp
//---------------------------------------------------------------------------
#include <vcl.h>
#pragma hdrstop

#include "DSACmp1.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "LbAsym"
#pragma link "LbCipher"
#pragma link "LbClass"
#pragma link "LbDSA"
#pragma resource "*.dfm"
TForm1 *Form1;
//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner)
    : TForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TForm1::btnGenParamsClick(TObject *Sender)
{
  Screen->Cursor = crHourGlass;
  StatusBar1->SimpleText = " Generating DSA parameters - this may take a while";

  try {
    if (LbDSA1->GeneratePQG()) {
      edtQ->Text = LbDSA1->PrivateKey->QAsString;
      mmoP->Lines->Text = LbDSA1->PrivateKey->PAsString;
      mmoG->Lines->Text = LbDSA1->PrivateKey->GAsString;
      StatusBar1->SimpleText = "";
    }
    else
      StatusBar1->SimpleText = " DSA parameter failure";
  }
  __finally {
    Screen->Cursor = crDefault;
  }
}
//---------------------------------------------------------------------------
void __fastcall TForm1::btnGenXYClick(TObject *Sender)
{
  Screen->Cursor = crHourGlass;
  StatusBar1->SimpleText = " Generating X and Y";

  try {
    LbDSA1->GenerateXY();
    edtX->Text = LbDSA1->PrivateKey->XAsString;
    mmoY->Lines->Text = LbDSA1->PublicKey->YAsString;
    StatusBar1->SimpleText = "";
  }
  __finally {
    Screen->Cursor = crDefault;
  }
}
//---------------------------------------------------------------------------
void __fastcall TForm1::btnSignClick(TObject *Sender)
{
  Screen->Cursor = crHourGlass;
  StatusBar1->SimpleText = " Signing message";

  try {
    LbDSA1->SignString(edtMsg->Text);
    edtR->Text = LbDSA1->SignatureR->IntStr;
    edtS->Text = LbDSA1->SignatureS->IntStr;
    StatusBar1->SimpleText = "";
  }
  __finally {
    Screen->Cursor = crDefault;
  }
}
//---------------------------------------------------------------------------
void __fastcall TForm1::btnVerifyClick(TObject *Sender)
{
  Screen->Cursor = crHourGlass;
  StatusBar1->SimpleText = " Verifying signature";

  try {
    if (LbDSA1->VerifyString(edtMsg->Text))
      StatusBar1->SimpleText = " Verification PASSED";
    else
      StatusBar1->SimpleText = " Verification FAILED";
  }
  __finally {
    Screen->Cursor = crDefault;
  }
}
//---------------------------------------------------------------------------
void __fastcall TForm1::btnSetParamsClick(TObject *Sender)
{
  LbDSA1->PrivateKey->QAsString = edtQ->Text;
  LbDSA1->PrivateKey->PAsString = mmoP->Lines->Text;
  LbDSA1->PrivateKey->GAsString = mmoG->Lines->Text;
  LbDSA1->PublicKey->QAsString = edtQ->Text;
  LbDSA1->PublicKey->PAsString = mmoP->Lines->Text;
  LbDSA1->PublicKey->GAsString = mmoG->Lines->Text;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::btnSetXYClick(TObject *Sender)
{
  LbDSA1->PrivateKey->XAsString = edtX->Text;
  LbDSA1->PublicKey->YAsString = mmoY->Text;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::LbDSA1GetSeed(TObject *Sender, TLbDSABlock &Block)
{
  HexToBuffer(edtSeed->Text, Block, sizeof(Block));
}
//---------------------------------------------------------------------------
void __fastcall TForm1::LbDSA1GetXKey(TObject *Sender, TLbDSABlock &Block)
{
  HexToBuffer(edtXKey->Text, Block, sizeof(Block));
}
//---------------------------------------------------------------------------
void __fastcall TForm1::LbDSA1GetKKey(TObject *Sender, TLbDSABlock &Block)
{
  HexToBuffer(edtKKey->Text, Block, sizeof(Block));
}
//---------------------------------------------------------------------------
void __fastcall TForm1::LbDSA1GetR(TObject *Sender, TLbDSABlock &Block)
{
  HexToBuffer(edtR->Text, Block, sizeof(Block));
}
//---------------------------------------------------------------------------
void __fastcall TForm1::LbDSA1GetS(TObject *Sender, TLbDSABlock &Block)
{
  HexToBuffer(edtS->Text, Block, sizeof(Block));
}
//---------------------------------------------------------------------------
void __fastcall TForm1::LbDSA1Progress(TObject *Sender, bool &Abort)
{
  Application->ProcessMessages();
  Abort = chkAbort->Checked;
}
//---------------------------------------------------------------------------
